coreos-ostree-importer/coreos_ostree_importer.py

```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/python3
+import fedora_messaging
+import fedora_messaging.api
+
+import json
+import logging
+import os
+import re
+import subprocess
+import sys
+import traceback
+import tempfile
+import os
+import tarfile
+import hashlib
+
+
+import botocore
+import boto3
+
+# from fedora_messaging import message
+# from fedora_messaging.api import publish
+# from fedora_messaging.config import conf
+
+# Set local logging
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+FEDORA_MESSAGING_TOPIC_LISTEN = (
+    "org.fedoraproject.prod.coreos.build.request.ostree-import"
+)
+FEDORA_MESSAGING_TOPIC_RESPOND = FEDORA_MESSAGING_TOPIC_LISTEN + ".finished"
+
+
+# We are processing the org.fedoraproject.prod.coreos.build.request.ostree-import topic
+# https://apps.fedoraproject.org/datagrepper/raw?topic=org.fedoraproject.prod.coreos.build.request.ostree-import&delta=100000
+EXAMPLE_MESSAGE_BODY = json.loads(
+    """
+{
+    "build_id": "30.20190905.0",
+    "stream": "testing",
+    "basearch": "x86_64",
+    "commit": "s3://fcos-builds/prod/streams/testing/builds/30.20190905.0/x86_64/ostree-commit.tar",
+    "checksum": "sha256:d01db6939e7387afa2492ac8e2591c53697fc21cf16785585f7f1ac0de692863",
+    "ostree_ref": "fedora/x86_64/coreos/testing",
+    "ostree_checksum": "b4beca154dab3696fd04f32ddab818102caa9247ec3192403adb9aaecc991bd9",
+    "target_repo": "prod"
+}
+"""
+)
+
+KNOWN_OSTREE_REPOS = {
+    "prod": "/mnt/koji/ostree/repo",
+    "compose": "/mnt/koji/compose/ostree/repo",
+}
+
+# Given a repo (and thus an input JSON) analyze existing koji tag set
+# and tag in any missing packages
+class Consumer(object):
+    def __init__(self):
+        logger.info(
+            "Processing messages with topic: %s" % FEDORA_MESSAGING_TOPIC_LISTEN
+        )
+
+    def __call__(self, message: fedora_messaging.api.Message):
+        # Catch any exceptions and don't raise them further because
```

Comment:
Hmm, I think a better approach here would be to define a new exception type and throw that in our code when checking for error conditions.

This kind of exception handling means that we don't discriminate between e.g. platform errors, programmer errors, and just invalid messages.

Reply:
This runs in an infinite loop in openshift. What I've found is that it's much easier to have the exceptions get thrown and viewable in the current container's log history than it is to distinguish between exceptions and `raise` or `pass` based on the type. 

If something isn't behaving properly then we can see the logs and fix it, right?

Reply:
Right, but a service being "up" (or Ready if there's a readiness probe) in Kubernetes should mean that it's available to handle requests. If in fact there's actually something wrong and we're not "passing that up" to Kubernetes, we lose those semantics and e.g. basic telemetry/alerts/`oc get pods` become less useful.

Of course, since we're pretty integrated right now, I agree it doesn't matter too much (if importing fails, we'll just go look at the container logs and see what's up). I tried looking for a good doc on best practices around error-handling for k8s services, but couldn't find anything good. Anyway, overall I don't think this is the best approach, but I'm good with just getting this in as is for now!

Reply:
I agree that what we're doing here is a bit of an anti-pattern and that errors should be bubbled up and reported at a higher level. AFAIU that higher level reporting/notifications doesn't exist yet (at least for this specific openshift environment) so this is the best I've come up with to make sure I can see the error messages and not lose them. Let's leave this as is for now and modify when the reporting situation is better?

Reply:
Yup, sure.

Reply:
👍